// @sumup/foundry/babel.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const merge = require('webpack-merge');

const withoutPreset = preset => arr => arr.filter(el => {
  const isStringPreset = typeof el === 'string' && el === preset;
  const isArrayPreset = el.constructor.name === 'Array' && el[0] === preset;
  return !isStringPreset && !isArrayPreset;
});
const overwriteEnvPreset = merge({
  customizeArray(a, b, key) {
    if (key === 'presets') {
      const aWithoutEnv = withoutPreset('@babel/preset-env')(a);
      return Object.values(merge(aWithoutEnv, b));
    }

    return undefined;
  }

});
const base = {
  presets: ['@babel/preset-env']
};
const react = {
  plugins: ['@babel/plugin-proposal-object-rest-spread', '@babel/plugin-proposal-class-properties'],
  presets: ['@babel/preset-react', {
    development: process.env.BABEL_ENV === 'development'
  }]
};
const node = overwriteEnvPreset(base, {
  // FIXME: this disables module transpilation in normal node envioonments. Needs
  //        to be fixed! potentially by editing this project's .babelrc.js accordingly.
  presets: [['@babel/preset-env', {
    targets: {
      node: '8.4.0'
    },
    modules: false
  }]]
});
const webpack = {
  plugins: ['@babel/plugin-syntax-dynamic-import'],
  presets: [['@babel/preset-env', {
    modules: false
  }]]
};
const lodash = {
  plugins: ['lodash']
};
const defaultReactSpa = overwriteEnvPreset(base, react, webpack, lodash);
const BABEL_CONFIGS = ['base', 'react', 'node', 'webpack', 'lodash'];

exports.withoutPreset = withoutPreset;
exports.overwriteEnvPreset = overwriteEnvPreset;
exports.base = base;
exports.react = react;
exports.node = node;
exports.webpack = webpack;
exports.lodash = lodash;
exports.defaultReactSpa = defaultReactSpa;
exports.BABEL_CONFIGS = BABEL_CONFIGS;
