// @sumup/foundry/cli.js
#!/usr/bin/env node
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var yargs = _interopDefault(require('yargs'));
var merge = _interopDefault(require('webpack-merge'));
var path = require('path');
var fs = require('fs');
var util = require('util');
var fp = require('lodash/fp');
var child_process = require('child_process');

const overwritePresets = merge({
  customizeArray(a, b, key) {
    return key === 'extends' ? b : undefined;
  }

});
const base = {
  extends: ['airbnb-base', 'plugin:jest/recommended', 'plugin:prettier/recommended'],
  plugins: ['prettier', 'jest'],
  rules: {
    'no-use-before-define': ['error', {
      functions: false
    }],
    'max-len': ['error', {
      code: 80,
      tabWidth: 2,
      ignoreComments: true,
      ignoreUrls: true
    }],
    curly: ['error', 'all'],
    'no-underscore-dangle': ['error', {
      allow: ['__DEV__', '__PRODUCTION__', '__TEST__']
    }]
  },
  globals: {
    __DEV__: true,
    __PRODUCTION__: true,
    __TEST__: true
  },
  parser: 'babel-eslint',
  parserOptions: {
    allowImportExportEverywhere: true,
    ecmaFeatures: {
      ecmaVersion: 2017,
      impliedStrict: true
    }
  },
  overrides: [{
    files: ['**/*spec.js'],
    rules: {
      'max-len': ['error', {
        code: 80,
        tabWidth: 2,
        ignorePattern: '^\\s*it(?:\\.(?:skip|only))?\\(',
        ignoreComments: true,
        ignoreUrls: true
      }]
    },
    globals: {
      mount: true,
      shallow: true,
      render: true,
      create: true,
      axe: true,
      renderToHtml: true
    },
    env: {
      'jest/globals': true
    }
  }]
}; // TODO: add node specific config here.

const node = overwritePresets(base, {
  env: {
    node: true
  }
});
const react = overwritePresets(base, {
  extends: ['airbnb-base', 'plugin:react/recommended', 'plugin:jsx-a11y/recommended', 'plugin:prettier/recommended', 'prettier/react'],
  plugins: ['react', 'jsx-a11y'],
  rules: {
    'import/no-extraneous-dependencies': ['error', {
      devDependencies: true
    }]
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true
    }
  },
  env: {
    browser: true
  }
});
const ESLINT_CONFIGS = ['base', 'node', 'react'];

var eslint = /*#__PURE__*/Object.freeze({
  overwritePresets: overwritePresets,
  base: base,
  node: node,
  react: react,
  ESLINT_CONFIGS: ESLINT_CONFIGS
});

const merge$1 = require('webpack-merge');

const withoutPreset = preset => arr => arr.filter(el => {
  const isStringPreset = typeof el === 'string' && el === preset;
  const isArrayPreset = el.constructor.name === 'Array' && el[0] === preset;
  return !isStringPreset && !isArrayPreset;
});
const overwriteEnvPreset = merge$1({
  customizeArray(a, b, key) {
    if (key === 'presets') {
      const aWithoutEnv = withoutPreset('@babel/preset-env')(a);
      return Object.values(merge$1(aWithoutEnv, b));
    }

    return undefined;
  }

});
const base$1 = {
  presets: ['@babel/preset-env']
};
const react$1 = {
  plugins: ['@babel/plugin-proposal-object-rest-spread', '@babel/plugin-proposal-class-properties'],
  presets: ['@babel/preset-react', {
    development: process.env.BABEL_ENV === 'development'
  }]
};
const node$1 = overwriteEnvPreset(base$1, {
  // FIXME: this disables module transpilation in normal node envioonments. Needs
  //        to be fixed! potentially by editing this project's .babelrc.js accordingly.
  presets: [['@babel/preset-env', {
    targets: {
      node: '8.4.0'
    },
    modules: false
  }]]
});
const webpack = {
  plugins: ['@babel/plugin-syntax-dynamic-import'],
  presets: [['@babel/preset-env', {
    modules: false
  }]]
};
const lodash = {
  plugins: ['lodash']
};
const defaultReactSpa = overwriteEnvPreset(base$1, react$1, webpack, lodash);
const BABEL_CONFIGS = ['base', 'react', 'node', 'webpack', 'lodash'];

var babel = /*#__PURE__*/Object.freeze({
  withoutPreset: withoutPreset,
  overwriteEnvPreset: overwriteEnvPreset,
  base: base$1,
  react: react$1,
  node: node$1,
  webpack: webpack,
  lodash: lodash,
  defaultReactSpa: defaultReactSpa,
  BABEL_CONFIGS: BABEL_CONFIGS
});

// eslint-disable-next-line import/prefer-default-export

var react$2 = (plop => {
  plop.setHelper('eq', (a, b) => a === b);
  plop.setHelper('not', (a, b) => a !== b);
  const pascalCase = plop.getHelper('pascalCase');
  const ERRORS = {
    INVALID_COMPONENT_TYPE: (type, supportedTypes) => [`We don't support "${type}" components, sorry.`, `You may try one of: ${pascalCase(supportedTypes.join(', '))}, instead.`],
    INVALID_FILE_TYPE: file => `We don't have templates for "${file}" files, 😞.`,
    INVALID_DESTINATION: path$$1 => `We couldn't find the destination folder ${path$$1} in your project, 🤷.`
  };
  const COMPONENT_TYPES = {
    STYLED: 'styled',
    FUNCTIONAL: 'functional',
    STATEFUL: 'stateful'
  };
  const COMPONENT_FILES = {
    COMPONENT: 'component',
    COMPONENT_SPEC: 'component-spec',
    STORY: 'story',
    SERVICE: 'service',
    SERVICE_SPEC: 'service-spec',
    INDEX: 'index'
  };
  const TEMPLATE_DIR = `${__dirname}/plop-templates`;

  const raiseErrorAndExit = message => {
    // eslint-disable-next-line no-console
    console.error([message, 'Please try again. 👋'].join(' '));
    process.exit(1);
  };

  const getComponentTemplateName = type => {
    switch (type) {
      case COMPONENT_TYPES.STYLED:
        return 'styled-component.hbs';

      case COMPONENT_TYPES.FUNCTIONAL:
        return 'functional-component.hbs';

      case COMPONENT_TYPES.STATEFUL:
        return 'stateful-component.hbs';

      default:
        return raiseErrorAndExit(ERRORS.INVALID_COMPONENT_TYPE);
    }
  };

  const getFileName = (file, name) => {
    const fileNameMap = {
      [COMPONENT_FILES.COMPONENT]: `${name}.js`,
      [COMPONENT_FILES.COMPONENT_SPEC]: `${name}.spec.js`,
      [COMPONENT_FILES.STORY]: `${name}.story.js`,
      [COMPONENT_FILES.SERVICE]: `${name}Service.js`,
      [COMPONENT_FILES.SERVICE_SPEC]: `${name}Service.spec.js`,
      [COMPONENT_FILES.INDEX]: 'index.js'
    };
    return fileNameMap[file];
  };
  /**
   * Generating React components and
   */


  plop.setGenerator('component', {
    description: 'React component',
    prompts: [// Component name
    {
      type: 'input',
      name: 'name',
      message: "What's the name of your component?"
    }, // Component type
    {
      type: 'list',
      name: 'type',
      message: 'What type of component do you need?',
      choices: [{
        name: 'Styled',
        value: COMPONENT_TYPES.STYLED
      }, {
        name: 'Functional',
        value: COMPONENT_TYPES.FUNCTIONAL
      }, {
        name: 'Stateful',
        value: COMPONENT_TYPES.STATEFUL
      }],
      default: 'functional'
    }, // Get the path
    {
      type: 'input',
      name: 'destinationPath',
      message: 'Where would you like to put your component?',
      default: path.relative(process.cwd(), `${plop.getPlopfilePath()}/src/components`)
    }, // Files
    {
      type: 'checkbox',
      name: 'files',
      message: 'Which files does your component need?',
      choices: [{
        name: 'Component',
        value: COMPONENT_FILES.COMPONENT,
        checked: true
      }, {
        name: 'Component spec',
        value: COMPONENT_FILES.COMPONENT_SPEC,
        checked: true
      }, {
        name: 'Story',
        value: COMPONENT_FILES.STORY,
        checked: false
      }, {
        name: 'Service',
        value: COMPONENT_FILES.SERVICE,
        checked: false
      }, {
        name: 'Service spec',
        value: COMPONENT_FILES.SERVICE_SPEC,
        checked: false
      }]
    }],
    actions: ({
      name,
      type,
      files,
      destinationPath
    }) => {
      const absDestinationPath = `${plop.getPlopfilePath()}/${destinationPath}`;
      const capitalizedName = pascalCase(name);
      const allFiles = files.includes(COMPONENT_FILES.COMPONENT) ? files.concat(COMPONENT_FILES.INDEX) : files;
      const actions = allFiles.reduce((acc, file) => {
        if (!Object.values(COMPONENT_FILES).includes(file)) {
          raiseErrorAndExit(ERRORS.INVALID_FILE);
        }

        const templateFileName = file === COMPONENT_FILES.COMPONENT ? getComponentTemplateName(type) : `${file}.hbs`;
        return acc.concat({
          type: 'add',
          path: path.join(absDestinationPath, capitalizedName, getFileName(file, name)),
          templateFile: path.join(TEMPLATE_DIR, 'react', templateFileName)
        });
      }, []);
      return actions;
    }
  });
});

var base$3 = (() => {});

/* istanbul ignore file */
const PLOP_CONFIGS = ['base', 'react'];

var plop = /*#__PURE__*/Object.freeze({
  PLOP_CONFIGS: PLOP_CONFIGS,
  react: react$2,
  base: base$3
});

const SUPPORTED_CONFIGS = ['eslint', 'babel', 'prettier', 'plop', 'husky', 'lint-staged'];

const writeFileAsync = util.promisify(fs.writeFile);
const {
  ESLINT_CONFIGS: ESLINT_CONFIGS$1
} = eslint;
const {
  BABEL_CONFIGS: BABEL_CONFIGS$1
} = babel;
const {
  PLOP_CONFIGS: PLOP_CONFIGS$1
} = plop;

function getConfigName(name, val) {
  const configs = {
    eslint: ESLINT_CONFIGS$1,
    prettier: ['base'],
    husky: ['base'],
    'lint-staged': ['base'],
    babel: BABEL_CONFIGS$1,
    plop: PLOP_CONFIGS$1
  };
  const supportedConfigs = fp.get(name, configs);
  const isSupportedConfig = fp.includes(val, supportedConfigs);

  if (!isSupportedConfig) {
    console.warn(`Config ${val} is not available for ${name}. Falling back to base config.`);
    return 'base';
  }

  return val;
}

function createConfigExport(name, config) {
  return `module.exports = require('@sumup/foundry/${name}').${config}`;
} // eslint-disable-next-line consistent-return


async function writeConfigFile(name, content, targetDir) {
  const filenames = {
    eslint: '.eslintrc.js',
    prettier: 'prettier.config.js',
    babel: '.babelrc.js',
    plop: 'plopfile.js',
    husky: '.huskyrc.js',
    'lint-staged': 'lint-staged.config.js'
  };
  const filename = fp.get(name, filenames);
  const path$$1 = path.resolve(targetDir, filename);

  try {
    return await writeFileAsync(path$$1, content);
  } catch (e) {
    console.error(`An error occured writing ${filename} to ${targetDir}.`);
    console.error(e);
    process.exit(1);
  }
}

const getConfigs = fp.flow(params => {
  const {
    all
  } = params;

  if (all) {
    return fp.zipObject(SUPPORTED_CONFIGS, Array.from(Array(SUPPORTED_CONFIGS.length).keys()).map(() => 'base'));
  }

  const picked = fp.flow(fp.pickAll(SUPPORTED_CONFIGS), fp.mapValues(v => v === true ? 'base' : v))(params);
  return picked;
}, fp.omitBy(val => !val));
function bootstrap(params) {
  // TODO: handle case where someone writes eslint config but not prettier
  //       config. Should at least get a message telling them they will either
  //       need to overwrite the eslint plugins/presets or provide one manually.
  const {
    targetDir
  } = params;
  const configParams = getConfigs(params);
  const tools = fp.keys(configParams);
  const configs = fp.reduce((acc, tool) => fp.assign(acc, {
    [tool]: getConfigName(tool, configParams[tool])
  }), {}, tools);
  const configExportStrings = fp.reduce((acc, tool) => ({ ...acc,
    [tool]: createConfigExport(tool, configs[tool])
  }), {}, tools); // TODO: Add a flag for the target directory.

  return fp.flow(fp.map(tool => writeConfigFile(tool, configExportStrings[tool], targetDir)), promises => Promise.all(promises) // breaks if you don't put the arrow function here.
  )(tools);
}

const DEFAULT_OPTIONS = {
  cwd: process.cwd(),
  detached: true,
  stdio: 'inherit'
};

function getBufferContent(chunks) {
  return Buffer.isBuffer(chunks[0]) ? Buffer.concat(chunks).toString('utf8') : null;
}

function asyncSpawn(cmd, args, options) {
  const stdout = [];
  return new Promise((resolve, reject) => {
    const child = child_process.spawn(cmd, args, { ...DEFAULT_OPTIONS,
      ...options
    });

    if (child.stdout) {
      child.stdout.on('data', chunk => {
        stdout.push(chunk);
      });
    }

    child.on('close', code => {
      if (code !== 0) {
        // eslint-disable-next-line no-console
        const err = new Error(`${cmd} exited with an error (code ${code}).`);
        err.log = getBufferContent(stdout);
        reject(err);
        return;
      }

      resolve(getBufferContent(stdout));
    });
    child.on('error', err => {
      reject(err);
    });
  });
}

const readFileAsync = util.promisify(fs.readFile);
const accessAsync = util.promisify(fs.access);

function shouldStopRecursion(path$$1) {
  return !/^.+node_modules/.test(path$$1);
}

async function resolveTo(path$$1, name) {
  if (shouldStopRecursion(path$$1)) {
    return null;
  }

  const packageJsonPath = path.join(path$$1, name);

  try {
    await accessAsync(packageJsonPath);
    return packageJsonPath;
  } catch (err) {
    const parentPath = path.resolve(path$$1, '..');
    return resolveTo(parentPath, name);
  }
}

async function getPackageJsonPath(name, useRelative = false) {
  const pathMain = require.resolve(name);

  const pathPackage = await resolveTo(pathMain, 'package.json');
  return useRelative ? path.relative(__dirname, pathPackage) : pathPackage;
}

function isRelativePath(path$$1) {
  const firstChar = path$$1.split()[0];
  return firstChar === '.';
}

async function loadJson(path$$1) {
  const isRelative = isRelativePath(path$$1);

  if (isRelative) {
    throw new TypeError(`Relative paths are not supported: ${path$$1}`);
  }

  try {
    const data = await readFileAsync(path$$1);
    return JSON.parse(data);
  } catch (err) {
    return new Error(`Path does not exist. ${path$$1}`);
  }
}

async function resolveBinaryPath(name, useRelative = false) {
  try {
    // This could potentially break, if the name of a binary (name) is different
    // from the name of the package.
    const packageJsonPath = await getPackageJsonPath(name, useRelative);
    const {
      bin: packageBin
    } = await loadJson(packageJsonPath);
    const binaryPath = fp.isString(packageBin) ? packageBin : packageBin[name];

    if (!binaryPath) {
      return null;
    }

    return path.resolve(path.dirname(packageJsonPath), binaryPath);
  } catch (err) {
    return null;
  }
}

function getToolArguments() {
  // The standard 2 indicating node binary, executing script, and
  // the run command and the tool argument.
  const SKIP_COUNT = 4;
  const processArgs = process.argv;
  return processArgs.slice(SKIP_COUNT);
}

async function executeBinary(path$$1, args) {
  return asyncSpawn(path$$1, args, {
    stdio: 'inherit'
  });
}

async function run({
  argv
}) {
  const {
    _: commandArgs
  } = argv;
  const [, tool] = commandArgs;
  const binPath = await resolveBinaryPath(tool);

  if (!binPath) {
    console.error(`No executable found for ${tool}`);
    process.exit(1);
  }

  const binArgs = getToolArguments();

  try {
    await executeBinary(binPath, binArgs);
  } catch (err) {
    process.exit(1);
  }
}

const {
  ESLINT_CONFIGS: ESLINT_CONFIGS$2
} = eslint;
const {
  BABEL_CONFIGS: BABEL_CONFIGS$2
} = babel; // eslint-disable-next-line

yargs.command('bootstrap-config', 'Set up custom configurations for babel, eslint, prettier, etc.', yrgs => yrgs.option('eslint', {
  desc: 'The eslint config to write.',
  choices: [true, ...ESLINT_CONFIGS$2],
  coerce: val => val === true ? 'base' : val
}).option('babel', {
  desc: 'The eslint config to write.',
  coerce: val => val === true ? 'base' : val,
  choices: [true, ...BABEL_CONFIGS$2]
}).option('prettier', {
  desc: 'Write the prettier config.',
  coerce: val => val === true ? 'base' : val
}).option('plop', {
  desc: 'Write the plop config.',
  coerce: val => val === true ? 'base' : val
}).option('husky', {
  desc: 'Write the husky config.',
  coerce: val => val === true ? 'base' : val
}).option('lint-staged', {
  desc: 'Write the lint-staged config.',
  coerce: val => val === true ? 'base' : val
}).option('targetDir', {
  default: process.cwd(),
  desc: 'Directory to write configs to.',
  type: 'string'
}).option('all', {
  default: false,
  desc: 'Write all base configurations.',
  type: 'boolean'
}), args => execute('bootstrap', args)).command('run <tool> [...tool options]', 'Run any of the bundled tools.', args => execute('run', args)).showHelpOnFail(true).demandCommand(1, '').help().version();
/**
 * Fallback, if an unsupported command was specified
 */

const registeredCommands = yargs.getCommandInstance().getCommands();
const currentCommand = yargs.argv._[0];

if (!registeredCommands.includes(currentCommand)) {
  yargs.showHelp();
  process.exit(1);
}

function execute(command, args) {
  const commands = {
    bootstrap,
    run
  };
  const commandFn = commands[command];
  commandFn(args);
}
